worker/host.go

```diff
@@ -209,73 +231,138 @@ func (h *hostClient) SyncAccount(ctx context.Context, rev *types.FileContractRev
-	settings, err := d.rhp4.Settings(ctx, d.hi.PublicKey, d.hi.V2SiamuxAddr())
+func (c *hostV2DownloadClient) Prices(ctx context.Context) (rhpv4.HostPrices, error) {
+	settings, err := c.rhp4.Settings(ctx, c.hi.PublicKey, c.hi.V2SiamuxAddr())
 	if err != nil {
 		return rhpv4.HostPrices{}, err
 	}
 	return settings.Prices, nil
 }
 
-func (d *hostV2DownloadClient) PublicKey() types.PublicKey { return d.hi.PublicKey }
+func (c *hostUploadClient) PriceTable(ctx context.Context, rev *types.FileContractRevision) (hpt api.HostPriceTable, cost types.Currency, err error) {
+	hpt, err = c.rhp3.PriceTable(ctx, c.hi.PublicKey, c.hi.SiamuxAddr, rhp3.PreparePriceTableAccountPayment(c.acc.Key()))
+	if err == nil {
+		cost = hpt.UpdatePriceTableCost
+	}
+	return
+}
+
+func (c *hostUploadClient) UploadSector(ctx context.Context, sectorRoot types.Hash256, sector *[rhpv2.SectorSize]byte) error {
+	rev, err := c.rhp3.Revision(ctx, c.fcid, c.hi.PublicKey, c.hi.SiamuxAddr)
+	if err != nil {
+		return fmt.Errorf("%w; %w", rhp3.ErrFailedToFetchRevision, err)
+	} else if rev.RevisionNumber == math.MaxUint64 {
+		return rhp3.ErrMaxRevisionReached
+	}
+
+	var hpt rhpv3.HostPriceTable
+	if err := c.acc.WithWithdrawal(func() (amount types.Currency, err error) {
+		pt, cost, err := c.pts.Fetch(ctx, c, nil)
+		if err != nil {
+			return types.ZeroCurrency, err
+		}
+		hpt = pt.HostPriceTable
+
+		gc, err := GougingCheckerFromContext(ctx)
+		if err != nil {
+			return cost, err
+		}
+		if breakdown := gc.CheckV1(nil, &pt.HostPriceTable); breakdown.Gouging() {
+			return cost, fmt.Errorf("%w: %v", gouging.ErrPriceTableGouging, breakdown)
+		}
+		return cost, nil
+	}); err != nil {
+		return err
+	}
+
+	cost, err := c.rhp3.AppendSector(ctx, sectorRoot, sector, &rev, c.hi.PublicKey, c.hi.SiamuxAddr, c.acc.ID(), hpt, c.rk)
+	if err != nil {
+		return fmt.Errorf("failed to upload sector: %w", err)
+	}
+
+	c.csr.RecordV1(rev, api.ContractSpending{Uploads: cost})
+	return nil
+}
+
+func (c *hostV2UploadClient) UploadSector(ctx context.Context, sectorRoot types.Hash256, sector *[rhpv2.SectorSize]byte) error {
+	fc, err := c.rhp4.LatestRevision(ctx, c.hi.PublicKey, c.hi.V2SiamuxAddr(), c.fcid)
+	if err != nil {
+		return err
+	}
+
+	rev := rhp.ContractRevision{
+		ID:       c.fcid,
+		Revision: fc,
+	}
+
+	return c.acc.WithWithdrawal(func() (types.Currency, error) {
+		prices, err := c.pts.Fetch(ctx, c)
+		if err != nil {
+			return types.ZeroCurrency, err
+		}
+
+		res, err := c.rhp4.WriteSector(ctx, c.hi.PublicKey, c.hi.V2SiamuxAddr(), prices, c.acc.Token(), NewReaderLen(sector[:]), rhpv2.SectorSize, api.BlocksPerDay*3)
+		if err != nil {
+			return types.ZeroCurrency, fmt.Errorf("failed to write sector: %w", err)
+		}
+		cost := res.Usage.RenterCost()
+
+		res2, err := c.rhp4.AppendSectors(ctx, c.hi.PublicKey, c.hi.V2SiamuxAddr(), prices, c.rk, rev, []types.Hash256{res.Root})
+		if err != nil {
+			return types.ZeroCurrency, fmt.Errorf("failed to write sector: %w", err)
```

Comment:
This isn't quite correct. If the first RPC succeeded, we need to return `cost` regardless.

Reply:
Thanks, overlooked that when I split it up.